### ISSUE TITLE
Slow text transfer/printing for Ipython

### ISSUE THREAD

**Issue by tblazina** [wuub/SublimeREPL#469]
I have the problem that when I evaluate something in SublimeREPL: Ipython, the text transfer is incredibly slow and often times causes Sublime Text 3 to crash. I do not know what I've changed but it previously did not have this problem.

For example, if I'm looking at a pandas dataframe object and do  data_frame.head(100), it will take >10 seconds to print. 

I do not have the same problem if I just open an ipython console through windows command line and run the same code. 

Anyone else seen this issue?

**Comment by djdjoko** [comment]
Yes I have a similar problem with SublimeREPL R. It is extremely slow printing data to screen even when data is relatively small.

**Comment by RSEm** [comment]
**Edit: Having tested this a bit, it seems to only work on windows. For some reason, the below changes break the repl on linux. I think the slow transfer problem is also windows-specific, though.**

Having recently dealt with a similar issue, I thought I would offer a possible solution.
**NOTE: This is a simple workaround. I'm not particularly familiar with sublime's plugin design. This change may cause issues that I haven't considered. Please backup the files before altering.**

I almost exclusively use the REPL as an interactive python editor and most of my debugging is done with print statements. It appears sublimeREPL buffers everything that it needs to print before updating the screen. This seems to be done one symbol at a time, and can cause performance issues with long strings. To change this functionality, go to the main plugin file (Packages\SublimeREPL\sublimerepl.py) and find the function handle_repl_output() in the ReplView class. Change it like this:

```
def handle_repl_output(self):
        """Returns new data from Repl and bool indicating if Repl is still
           working"""
        if self.repl.apiv2:
            try:
                while True:
                    packet = self._repl_reader.queue.get_nowait()
                    if packet is None:
                        return False

                    self.handle_repl_packet(packet)

            except queue.Empty:
                return True

        else:
            try:
                packet = self._repl_reader.queue.get_nowait()
                if packet is None:
                    return False
                for _ in xrange(1000):
                    try:
                        packet += self._repl_reader.queue.get_nowait()
                    except queue.Empty: break

                self.handle_repl_packet(packet)
                return True

            except queue.Empty:
                return True
```

I used the same check with self.repl.apiv2 as in handle_repl_packet() to ensure that the new functionality is only applied to the packets that are sent directly to write(), which I believe are just unicode strings.
The loop strings together multiple packets and sends them all to write() as one long string. The repl image is updates when you leave the function, so the length of the loop determines how many symbols are printed per cycle. You can increase this number for faster printing. You may also want to reduce the timeout in update_view_loop() at the same time (minimum value is 1.).

**Comment by oro1989** [comment]
@RSEm your solution worked for infinite loops hanging and loops that use print taking long time.Thanks

**Comment by ghost** [comment]
@RSEm Thanks, this works for me too!  You saved me minutes of waiting for so many cases, it went from 5 minutes to 5 seconds!

**Comment by agyurov** [comment]
brilliant!

**Comment by listerma** [comment]
@RSEm  Very very very grateful to your idea,it works very well ,great help!     windows 10  V3 BUILD3143

**Comment by Fede112** [comment]
Hi, I think I'm having this same problem with SublimeREPL but I use Linux. Does anyone know how to fix this for Linux? I will try to fix it myself but I'm kind of a newbie so...

**Comment by mew1033** [comment]
Any chance we can see this integrated in directly?

Edit: Actually, I couldn't get it to work. I just made the REPL hang when starting. Running Windows 10.

**Comment by JohnRainer** [comment]
## Two errors raised and a bug：
- global name 'xrange' is not defind  
- TypeError("Can't convert 'NoneType' object to str implicitly",)
- pdb.set_trace() will lead to endless loop

##  'xrange' is not defind  
solution: use the range instead of xrange

## TypeError
That is beacause a string was concatenated with the NoneType in
`packet += self._repl_reader.queue.get_nowait()`

I made som modifications

```
def handle_repl_output(self):
        """Returns new data from Repl and bool indicating if Repl is still
           working"""
        if self.repl.apiv2:
            try:
                while True:
                    packet = self._repl_reader.queue.get_nowait()
                    if packet is None:
                        return False
                    self.handle_repl_packet(packet)

            except queue.Empty:
                return True

        else:
            try:
                packet = self._repl_reader.queue.get_nowait()
                if packet is None:
                    return False
                for _ in range(1000):
                    new_packet = self._repl_reader.queue.get_nowait()
                    if new_packet is None:
                        break
                    else:
                        packet += new_packet
                    

                self.handle_repl_packet(packet)
                return False

            except queue.Empty:
                return True
```

## pdb endless loop

pdb.set_trace() will lead to endless loop
i guass the problem was caused by packet combination,  but don't know how to deal with it
```
import pdb
print("hello world")
pdb.set_trace()
```

Test in windows 10 version 1803，looking forward to your help.

**Comment by RSEm** [comment]
Upgraded to python 3.6 and sublime text 3 recently and got a chance to revisit this. Replacing xrange() with range() should be the only necessary tweak. I haven't personally experienced further issues with my original solution.

@JohnRainer 
The reason for your TypeError seems to be that you missed the try-except that catches queue.Empty inside the loop.
This will also make the repl hang because get_nowait() throws the error and kicks you back to the outer try-except, skipping the call to handle_repl_packet() completely.

**Comment by Erimus-Koo** [comment]
Windows 10, sublime text v3.1.1, python 3.7.  
I tried @RSEm 's code, but it looks infinite loop at the last part.  
So I modified it like below, and now it works good for me.  
Hope to be helpful.
```
    def handle_repl_output(self):
        """Returns new data from Repl and bool indicating if Repl is still
           working"""
        # try:
        #     while True:
        #         packet = self._repl_reader.queue.get_nowait()
        #         if packet is None:
        #             return False

        #         self.handle_repl_packet(packet)

        # except queue.Empty:
        #     return True

        if self.repl.apiv2:
            try:
                while True:
                    packet = self._repl_reader.queue.get_nowait()
                    if packet is None:
                        return False

                    self.handle_repl_packet(packet)

            except queue.Empty:
                return True

        else:
            try:
                packet = self._repl_reader.queue.get_nowait()
                for _ in range(999):
                    try:
                        packet += self._repl_reader.queue.get_nowait()
                    except Exception:
                        self.handle_repl_packet(packet)  # if break here, the repl will not close.
                        return False

                self.handle_repl_packet(packet)
                return True

            except queue.Empty:
                return True
```
**Warning**: I found it still has some problem, about the 'Exception' part, maybe stop printing accidentally. But if use 'queue.Empty' here, it will cause infinite loop. 'TypeError' is not good either.  

And I do also tried @JohnRainer 's method, but still possible to fall into an infinite loop.
```
        if self.repl.apiv2:
            try:
                while True:
                    packet = self._repl_reader.queue.get_nowait()
                    if packet is None:
                        return False

                    self.handle_repl_packet(packet)

            except queue.Empty:
                return True

        else:
            try:
                packet = self._repl_reader.queue.get_nowait()
                if packet is None:
                    return False

                for _ in range(999):
                    new_packet = self._repl_reader.queue.get_nowait()
                    if new_packet is None:
                        self.handle_repl_packet(packet)
                        return False
                    else:
                        packet += new_packet

                self.handle_repl_packet(packet)
                return True

            except queue.Empty:
                return True
```

---

I found the reason, if the printing be paused for a while, the repl will close.
```
print('abc')    # print ok
time.sleep(3)   # this step will stop repl
print('123)     # will not print
```
I have not thought about how to fix it.

**Comment by weidongzhou1994** [comment]
I find a bug: when I use the above code, the IPython can't tun in SublimeRepl.

**Comment by renato-ll** [comment]
> ## Two errors raised and a bug：
> * global name 'xrange' is not defind
> * TypeError("Can't convert 'NoneType' object to str implicitly",)
> * pdb.set_trace() will lead to endless loop
> 
> ## 'xrange' is not defind
> solution: use the range instead of xrange
> 
> ## TypeError
> That is beacause a string was concatenated with the NoneType in
> `packet += self._repl_reader.queue.get_nowait()`
> 
> I made som modifications
> 
> ```
> def handle_repl_output(self):
>         """Returns new data from Repl and bool indicating if Repl is still
>            working"""
>         if self.repl.apiv2:
>             try:
>                 while True:
>                     packet = self._repl_reader.queue.get_nowait()
>                     if packet is None:
>                         return False
>                     self.handle_repl_packet(packet)
> 
>             except queue.Empty:
>                 return True
> 
>         else:
>             try:
>                 packet = self._repl_reader.queue.get_nowait()
>                 if packet is None:
>                     return False
>                 for _ in range(1000):
>                     new_packet = self._repl_reader.queue.get_nowait()
>                     if new_packet is None:
>                         break
>                     else:
>                         packet += new_packet
>                     
> 
>                 self.handle_repl_packet(packet)
>                 return False
> 
>             except queue.Empty:
>                 return True
> ```
> 
> ## pdb endless loop
> pdb.set_trace() will lead to endless loop
> i guass the problem was caused by packet combination, but don't know how to deal with it
> 
> ```
> import pdb
> print("hello world")
> pdb.set_trace()
> ```
> 
> Test in windows 10 version 1803，looking forward to your help.

How to use it for persistent console?
Im using "i" tag to open python, so it doesnt close after i open

["python", "-i", "-u", "-m", "pdb", "$file_basename"]

But using your code it closes after the script runs
